Mathlib/CategoryTheory/Sites/Coherent.lean

```diff
@@ -82,4 +83,148 @@ lemma isSheaf_coherent [Precoherent C] (P : Cᵒᵖ ⥤ Type w) :
+  have hx_ext := Presieve.FamilyOfElements.Compatible.sieveExtend hx
+  let S := Sieve.generate (Presieve.ofArrows Y π)
+  have := (Sieve.forallYonedaIsSheaf_iff_colimit S).mpr ⟨h_colim⟩ W x_ext hx_ext
+  rcases this with ⟨t, t_amalg, t_uniq⟩
+  use t
+  constructor
+  · convert Presieve.isAmalgamation_restrict (Sieve.le_generate (Presieve.ofArrows Y π)) _ _ t_amalg
+    refine Eq.symm (Presieve.restrict_extend hx)
+  · exact fun y hy => t_uniq y <| Presieve.isAmalgamation_sieveExtend x y hy
+
+/-- The coherent topology on a precoherent category is subcanonical. -/
+theorem isSubcanonical : Sheaf.Subcanonical (coherentTopology C) :=
+  Sheaf.Subcanonical.of_yoneda_isSheaf _ isSheaf_yoneda_obj
+
+end coherentTopology
+
+variable {C : Type _} [Category C] [Precoherent C]
+
+variable {X : C}
+
+/--
+Effective epi families in a precoherent category are transitive, in the sense that an
+`EffectiveEpiFamily` and an `EffectiveEpiFamily` over each member, the composition is an
+`EffectiveEpiFamily`.
+Note: The finiteness condition is an artifact of the proof and is probably unneccessary
+-/
+theorem EffectiveEpiFamily.transitive_of_finite {α : Type} [Fintype α] {Y : α → C}
+    (π : (a : α) → (Y a ⟶ X)) (h : EffectiveEpiFamily Y π) {β : α → Type} [∀ (a: α), Fintype (β a)]
+    {Y_n : (a : α) → β a → C} (π_n : (a : α) → (b : β a) → (Y_n a b ⟶ Y a))
+    (H : ∀ a, EffectiveEpiFamily (Y_n a) (π_n a)) :
+EffectiveEpiFamily (fun (c : Σ a, β a) => Y_n c.fst c.snd) (fun c => π_n c.fst c.snd ≫ π c.fst)
+    := by
+  rw [← Sieve.effectiveEpimorphic_family]
+  suffices h₂ : (Sieve.generate (Presieve.ofArrows (fun (⟨a, b⟩ : Σ _, β _) => Y_n a b)
+        (fun ⟨a,b⟩ => π_n a b ≫ π a))) ∈ GrothendieckTopology.sieves (coherentTopology C) X by
+    change Nonempty _
+    rw [← Sieve.forallYonedaIsSheaf_iff_colimit]
+    exact fun W => coherentTopology.isSheaf_yoneda_obj W _ h₂
+  let h' := h
+  rw [← Sieve.effectiveEpimorphic_family] at h'
+  let H' := H
+  conv at H' =>
+    intro a
+    rw [← Sieve.effectiveEpimorphic_family]
+  -- Show that a covering sieve is a colimit, which implies the original set of arrows is regular
+  -- epimorphic. We use the transitivity property of saturation
+  apply Coverage.saturate.transitive X (Sieve.generate (Presieve.ofArrows Y π))
+  · apply Coverage.saturate.of
+    use α, inferInstance, Y, π
+    simp only [true_and]
+    exact Iff.mp (Sieve.effectiveEpimorphic_family Y π) h'
+  · intro V f ⟨Y₁, h, g, ⟨hY, hf⟩⟩
+    rw [← hf, Sieve.pullback_comp]
+    apply (coherentTopology C).pullback_stable'
+    apply coherentTopology.mem_sieves_of_hasEffectiveEpiFamily
+    -- Need to show that the pullback of the family `π_n` to a given `Y i` is effective epimorphic
+    rcases hY with ⟨i⟩
+    use β i, inferInstance, Y_n i, π_n i
+    constructor
+    · exact H i
+    · intro b
+      use Y_n i b, (𝟙 _), π_n i b ≫ π i
+      constructor
+      · exact ⟨(⟨i, b⟩ : Σ (i : α), β i)⟩
+      · exact Category.id_comp (π_n i b ≫ π i)
```

Comment:
```suggestion
    use β i, inferInstance, Y_n i, π_n i, H i 
    intro b
    use Y_n i b, (𝟙 _), π_n i b ≫ π i, ⟨(⟨i, b⟩ : Σ (i : α), β i)⟩
    exact Category.id_comp (π_n i b ≫ π i)
```
`use a,b` is just `refine \<a,b,?_\>` and `constructor is just `refine \<_,_,...,_\>`.